### ISSUE TITLE
Snapcraft-style version metadata handling

### ISSUE THREAD

**Issue by devec0** [canonical/charmcraft#159]
One current challenge with diagnosing and filing bugs for charms is knowing where in the source repository to start bisecting or researching, as there is no tight coupling between DVCS metadata and publish charm revisions. This gets a lot easier with the potential introduction of charm channels (similar to how snaps handle various release streams), assuming teams follow the popular workflow of having their stable charm releases mapped to per-stable channels, and having that workflow mirrored in their workflow in the form of branches, however there is still some guesswork here and an assumption about a given team's workflow.

What would be great, is if the version of a built charm could be set automatically (or at least, optionally, ala `version: git` from snapcraft) from DVCS metadata (for example, from the output of `git describe`) so that the current git tag or commit hash could be captured as part of the version pushed to the charm store. I'm deliberately trying not to be prescriptive about the how, and instead trying to frame the why. This would help hugely to disambiguate commits from published charms and really tighten the feedback loop for charm users when reporting bugs, and for charm developers when diagnosing reported issues with their charms. The actual implementation could achieve this goal in a few different ways. I mention the snapcraft workflow only because it's one that I personally find really ergonomic.

Thanks for all of your work on charmcraft!

**Comment by chipaca** [comment]
when you described this in chat I thought it was more than #37, but from your description above it's not that clear.
Do you want a single 'version' string for the charm itself, or a series of version information for everything used to build or otherwise included in the charm, a la manifest.yaml from snapcraft?

**Comment by devec0** [comment]
So, reading #37, it refers to the 'version' file. This file has taken many forms across different charms, but is kind of a workaround for the lack of first-class support for more detailed version information in the store and in the charm tooling.

Ideally, I would like to have the DVCS information at time of build stored and uploaded alongside each charm revision, using rich metadata that is supported by the store API and charmcraft. Discussing this with @jlounder - it was suggested this could be encoded into the existing version field, similar to how `snapcraft` can be configured via `snapcraft.yaml` into using different sources for automatic versioning. When `version: git` is used for example, it generates a long version string that includes the commit hash when there is no direct annotated tag to refer to the current commit, which would achieve the same goal.

So, in short, the technical goal is the same as #37, however the suggestion/request here is to actually make it part of the tooling and workflow to have this information user-controllable, if required, but default to using useful DVCS information for automatically calculating the version of a charm revision.

The end user in this case would see much nicer looking versions in the store, which could be used directly to find which tag or commit a specific build charm, on the store, was built from. Hope that clears up the intention some.

**Comment by facundobatista** [comment]
We'll introduce a version in `charmcraft.yaml` (that will end up in the file) and establish the mechanism `craftctl set-version $(do-whatever)` so the developer will have absolute control on how to set it. This way all needs will be addressed.

**Comment by facundobatista** [comment]
Closing this as we're actually moving away from the concept of "version" (see [this post in other issue for longer explanation](https://github.com/canonical/charmcraft/issues/37#issuecomment-1141078396)).
